contributors/design-proposals/customresources-validation.md

```diff
@@ -0,0 +1,493 @@
+
+### Outlook Admission Webhook
+
+Apiextensions-apiserver uses the normal REST endpoint implementation and only customizes the registry and the codecs. The admission plugins are inherited from the kube-apiserver (when running inside of it via apiserver delegation) and therefore they are supposed to apply to CRs as well.
+
+The webhook admission prototyped at https://github.com/kubernetes/kubernetes/pull/46316 is expected to work with CRs out of the box. Hence, for more advanced validation the webhook admission is an option as well (when it is merged).
+
+Clearly, providing such a webhook is more complex and therefore cannot replace a built-in JSON-Schema validation and its use-cases.
+
+## Implementation Plan
+
+The implementation is planned in the following steps:
+
+1. Add the proposed types to the v1beta1<sup id="f3">[3](#footnote3)</sup> version of the CRD type.
+2. Add a validation step to the CREATE and UPDATE REST handlers of the apiextensions-apiserver.
+ 
+As an optional follow-up, we can implement the OpenAPI part and with that enable client-side validation:
+
+3. Export the JSON-Schema via a dynamically served OpenAPI spec.
+
+## Appendix
+
+### Expressiveness of JSON-Schema
+
+The following example properties cannot be expressed using JSON-Schema:
+1. “In a PodSpec, for each `spec.Containers[*].volumeMounts[*].Name` there must be a `spec.Volumes[*].Name`”
+2. “The volume names in `PodSpec.Volumes` are unique” (`uniqueItems` only compares the complete objects, it cannot compare by key)
+ 
+Different versions within one CRD with a custom version field (i.e. not the one in apiVersion) **can** be expressed:
+
+```json
+{
+    "$schema": "http://json-schema.org/draft-04/schema#",
+    "title": "child_schema",
+    "type": "object",
+    "anyOf": [
+        {
+            "properties": {
+                "version": {
+                    "type": "string",
+                    "pattern": "^a$"
+                },
+                "spec": {
+                    "type": "object",
+                    "properties": {
+                        "foo": {}
+                    },
+                    "additionalProperties": false,
+                },
+            }
+        },
+        {
+            "properties": {
+                "version": {
+                    "type": "string",
+                    "pattern": "^b$"
+                },
+                "spec": {
+                    "type": "object",
+                    "properties": {
+                        "bar": {}
+                    },
+                    "additionalProperties": false,
+                },
+            }
+        }
+    ],
+}
+```
+
+This validates: 
+* `{"version": "a",  "spec": {"foo": 42}}`
+* `{"version": "b",  "spec": {"bar": 42}}`
+
+but not:
+* `{"version": "a",  "spec": {"bar": 42}}`.
+ 
+Note: this is a workaround while we do not support multiple versions and conversion for custom resources.
+
+### JSON-Schema Validation Runtime Complexity
```

Comment:
In general, it seems like this will be slower than native validation. We should consult with @kubernetes/sig-scalability-proprosals to define the performance bar this needs to meet and create regression tests for large numbers of objects.